compiler/compiler.js

```diff
@@ -0,0 +1,410 @@
+  const emitTime = ts.performance.getDuration("Emit");
+  stats.push(["Parse time", programTime]);
+  stats.push(["Bind time", bindTime]);
+  stats.push(["Check time", checkTime]);
+  stats.push(["Emit time", emitTime]);
+  stats.push(["Total TS time", programTime + bindTime + checkTime + emitTime]);
+  ts.performance.disable();
+  stats.push(["Duration", Date.now() - start]);
+}
+
+/**
+ * @typedef {object} Source
+ * @property {string} data
+ * @property {string=} hash
+ */
+
+/**
+ * @typedef {object} CompileRequest
+ * @property {Record<string, any>} compilerOptions
+ * @property {boolean} debug
+ * @property {string[]} rootNames
+ */
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+
+/**
+ * @param {CompileRequest} request
+ */
+function compile(request) {
+  const start = Date.now();
+  ts.performance.enable();
+  const { compilerOptions, debug, rootNames } = request;
+  debugFlag = debug;
+
+  println(">>> compile");
+
+  const { options, diagnostics: configFileParsingDiagnostics } = configure(
+    compilerOptions,
+  );
+
+  assert(rootNames.length === 1, "only single root names supported");
+  const [rootSpecifier] = rootNames;
+  const builderProgram = ts.createIncrementalProgram({
+    rootNames: [rootSpecifier],
+    options,
+    host,
+    configFileParsingDiagnostics,
+  });
+
+  const diagnostics = [
+    ...builderProgram.getConfigFileParsingDiagnostics(),
+    ...builderProgram.getSyntacticDiagnostics(),
+    ...builderProgram.getOptionsDiagnostics(),
+    ...builderProgram.getGlobalDiagnostics(),
+    ...builderProgram.getSemanticDiagnostics(),
+  ].filter(({ code }) => !IGNORED_DIAGNOSTICS.includes(code));
+
+  /** @type {ts.EmitResult} */
+  let emitResult;
+
+  if (diagnostics.length) {
+    emitResult = { emitSkipped: true, diagnostics };
+  } else {
+    emitResult = builderProgram.emit();
+  }
+
+  /** @type {[string, number][]} */
+  const stats = [];
+  const program = builderProgram.getProgram();
+  stats.push(["Files", program.getSourceFiles().length]);
+  stats.push(["Nodes", program.getNodeCount()]);
+  stats.push(["Identifiers", program.getIdentifierCount()]);
+  stats.push(["Symbols", program.getSymbolCount()]);
+  stats.push(["Types", program.getTypeCount()]);
+  stats.push(["Instantiations", program.getInstantiationCount()]);
+  addTimes(stats, start);
+  ts.performance.disable();
+
+  dispatch(
+    "op_set_emit_result",
+    Object.assign(emitResult, {
+      diagnostics: processDiagnostics(emitResult.diagnostics),
+      stats,
+    }),
+  );
+  println("<<< compile");
+}
+
+/**
+ * @typedef {object} TranspileRequest
+ * @property {Record<string, any>} compilerOptions
+ * @property {boolean} debug
+ * @property {Record<string, string>} sources
+ */
+
+/**
+ * @param {TranspileRequest} request
+ */
+function transpile(request) {
+  const start = Date.now();
+  ts.performance.enable();
+  const { compilerOptions, debug, sources } = request;
+  debugFlag = debug;
```

Comment:
I'm a bit lost here... Why is this method needed? Don't we want to do the transpiling using `swc` in Rust?

Reply:
Okay, I see that you mentioned it needs to be done in your comment https://github.com/denoland/deno/pull/6894#issuecomment-674349864